app/uk/gov/hmrc/gform/formtemplate/FormTemplateValidator.scala

```diff
@@ -95,6 +96,106 @@ object FormTemplateValidator {
+      .collect {
+        case s: Section.NonRepeatingPage => s.page.allFormComponentIds.toSet
+        case s: Section.RepeatingPage    => s.page.allFormComponentIds.toSet
+        case s: Section.AddToList        => (s.addAnotherQuestion.id :: s.pages.toList.flatMap(_.allFormComponentIds)).toSet
+      }
+      .toSet
+      .flatten
+
+    val allExprs: List[ExprWithPath] = FormTemplate.leafExprs.exprs(TemplatePath.root, formTemplate)
+
+    def invalid(path: TemplatePath, formComponentId: FormComponentId): Invalid =
+      Invalid(s"${path.path}: $formComponentId doesn't exist in the form")
+
+    allExprs.flatMap(_.referenceInfos).foldMap {
+      case ReferenceInfo.FormCtxExpr(path, FormCtx(formComponentId)) if !allFcIds(formComponentId) =>
+        invalid(path, formComponentId)
+      case ReferenceInfo.SumExpr(path, Sum(FormCtx(formComponentId))) if !allFcIds(formComponentId) =>
+        invalid(path, formComponentId)
+      case ReferenceInfo.CountExpr(path, Count(formComponentId)) if !allFcIds(formComponentId) =>
+        invalid(path, formComponentId)
+      case _ => Valid
+    }
+  }
+
+  def validateReferencesConstraints(formTemplate: FormTemplate): ValidationResult = {
+
+    val addToListIds: Set[FormComponentId] = formTemplate.sections
+      .collect { case s: Section.AddToList =>
+        s.pages.toList.flatMap(_.allFormComponentIds).toSet
+      }
+      .toSet
+      .flatten
+
+    val repeatedSectionIds: Set[FormComponentId] = formTemplate.sections
+      .collect { case s: Section.RepeatingPage =>
+        s.page.allFormComponentIds
+      }
+      .toSet
+      .flatten
+
+    val groupFieldsIds: Set[FormComponentId] = formTemplate.sections
+      .collect { case s: Section.NonRepeatingPage =>
+        s.page.fields.collect { case fc @ IsGroup(g) =>
+          g.fields.map(_.id)
+        }.flatten
+      }
+      .toSet
+      .flatten
+
+    val formTemplateExprs = FormTemplate.leafExprsNoSections.exprs(TemplatePath.root, formTemplate)
+
+    val withoutAddToLists: List[ExprWithPath] = formTemplate.sections.collect {
+      case s: Section.NonRepeatingPage => LeafExpr(TemplatePath.root, s)
+      case s: Section.RepeatingPage    => LeafExpr(TemplatePath.root, s)
+    }.flatten ++ formTemplateExprs
+
+    val withoutRepeatingPages: List[ExprWithPath] = formTemplate.sections.collect {
+      case s: Section.NonRepeatingPage => LeafExpr(TemplatePath.root, s)
+      case s: Section.AddToList        => LeafExpr(TemplatePath.root, s)
+    }.flatten ++ formTemplateExprs
+
+    val withoutGroups: List[ExprWithPath] = formTemplate.sections.collect {
+      case s: Section.NonRepeatingPage =>
+        Page.leafExprsNoFields.exprs(TemplatePath.root, s.page) ++
+          s.page.fields.flatMap {
+            case IsGroup(group) => Nil // We want to ignore groups
+            case otherwise      => LeafExpr(TemplatePath("fields"), otherwise)
+          }
+      case s: Section.RepeatingPage => LeafExpr(TemplatePath.root, s)
```

Comment:
what about groups in repeating page and add to list?

Reply:
Group can't exists in repeating page nor in add to list.

Reply:
Good to know, thanks